state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -957,8 +1026,117 @@ pub mod pallet {
 					debug_assert!(false, "Swap is not completed yet!");
 				}
 			}
+		}
 
-			Ok(())
+		fn process_swaps_for_block(
+			current_block: BlockNumberFor<T>,
+			block_to_process: BlockNumberFor<T>,
+		) -> Result<(), ()> {
+			let mut swaps_to_execute = SwapQueue::<T>::take(block_to_process);
+
+			loop {
+				if swaps_to_execute.is_empty() {
+					return Ok(())
+				}
+
+				match Self::execute_batch(swaps_to_execute.clone()) {
+					Ok(successful_swaps) => {
+						Self::process_swap_outcomes(&successful_swaps);
+						// Nothing to do here, all swaps are processed for block
+						return Ok(())
+					},
+					Err(err) => {
+						// Depending on the error, split the swaps into "satisfactory"
+						// (have a chance of succeeding if retried immediately), and "failed"
+						// (unlikely to succeed now and should be retried later or refunded).
+						let (satisfactory_swaps, failed_swaps) = {
+							match err {
+								BatchExecutionError::PriceLimitHit {
+									successful_swaps,
+									failed_swaps,
+								} => (successful_swaps, failed_swaps),
+								BatchExecutionError::SwapLegFailed { asset, direction, amount } => {
+									Self::deposit_event(Event::<T>::BatchSwapFailed {
+										asset,
+										direction,
+										amount,
+									});
+									(vec![], swaps_to_execute)
+								},
+								BatchExecutionError::DispatchError { error } => {
+									log::error!(
+										"Failed to execute swap batch at block {:?}: {:?}",
+										block_to_process,
+										error
+									);
+									(vec![], swaps_to_execute)
+								},
+							}
+						};
+
+						let retry_block = current_block + SWAP_RETRY_DELAY_BLOCKS.into();
+
+						for swap in failed_swaps {
+							dbg!(&swap.refund_params, retry_block);
+
+							match swap.refund_params {
+								Some(params)
+									if BlockNumberFor::<T>::from(params.refund_block) <
+										retry_block =>
+								{
+									// Reached refund block, schedule refund:
+									match T::EgressHandler::schedule_egress(
```

Comment:
Just want to mention that I think refunding should happen independently of the block processing / swapping logic. 

For example, for DCA, we might want to refund only the remainder of a swap, and transfer the rest.

I can also imagine a future scenario where we want fok-like behaviour on internal swaps, and *don't* need to refund via egress, but simply credit the funds back to the orginator's free balance. 

This comes back to the design problem of separating the swapping from the intent of the swap (is it internal/external/ccm/contract) from the execution strategy (dca, price limits, time limits). Here we've buried an external concern (what to do if the swap expires) inside a loop in an inner pallet function. This doesn't feel right.